### PR TITLE
schedule: fix datarace in `operator.check`

### LINKED ISSUE #8263: schedule: `operator.check` will meet datarace
### What did you do?

run a server which built with `-race`

### What did you expect to see?

no datarace

### What did you see instead?

```
WARNING: DATA RACE
Read at 0x00c059c28fc8 by goroutine 2253766:
  github.com/tikv/pd/pkg/schedule/operator.(*Operator).Check()
      /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/pd/pkg/schedule/operator/operator.go:382 +0x544
  github.com/tikv/pd/pkg/schedule/operator.(*Controller).pollNeedDispatchRegion()
      /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/pd/pkg/schedule/operator/operator_controller.go:262 +0x144
  github.com/tikv/pd/pkg/schedule/operator.(*Controller).PushOperators()
      /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/pd/pkg/schedule/operator/operator_controller.go:281 +0x2e
  github.com/tikv/pd/pkg/schedule.(*Coordinator).drivePushOperator()
      /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/pd/pkg/schedule/coordinator.go:412 +0x224
  github.com/tikv/pd/pkg/schedule.(*Coordinator).Run.func5()
      /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/pd/pkg/schedule/coordinator.go:491 +0x33

Previous write at 0x00c059c28fc8 by goroutine 37567:
  sync/atomic.CompareAndSwapInt64()
      /usr/local/go/src/runtime/race_amd64.s:316 +0xb
  sync/atomic.CompareAndSwapInt64()
      <autogenerated>:1 +0x1a
  github.com/tikv/pd/pkg/schedule/operator.(*Controller).Dispatch()
      /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/pd/pkg/schedule/operator/operator_controller.go:145 +0x104
  github.com/tikv/pd/server/cluster.(*RaftCluster).HandleRegionHeartbeat()
      /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/pd/server/cluster/cluster_worker.go:66 +0x5c6  github.com/tikv/pd/server.(*GrpcServer).RegionHeartbeat()
      /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/pd/server/grpc_service.go:1296 +0x1c59
  github.com/pingcap/kvproto/pkg/pdpb._PD_RegionHeartbeat_Handler()
      /go/pkg/mod/github.com/pingcap/kvproto@v0.0.0-20240403065636-c699538f7aa1/pkg/pdpb/pdpb.pb.go:9777 +0xb8
  github.com/grpc-ecosystem/go-grpc-prometheus.init.(*ServerMetrics).StreamServerInterceptor.func4()
      /go/pkg/mod/github.com/grpc-ecosystem/go-grpc-prometheus@v1.2.0/server_metrics.go:121 +0x15e
  go.etcd.io/etcd/etcdserver/api/v3rpc.Server.ChainStreamServer.func9.1()
      /go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.4.0/chain.go:83 +0xc3
  go.etcd.io/etcd/etcdserver/api/v3rpc.newStreamInterceptor.func1()
      /go/pkg/mod/go.etcd.io/etcd@v0.5.0-alpha.5.0.20240320135013-950cd5fbe6ca/etcdserver/api/v3rpc/interceptor.go:238 +0x6f0
  go.etcd.io/etcd/etcdserver/api/v3rpc.Server.ChainStreamServer.func9()
      /go/pkg/mod/github.com/grpc-ecosystem/go-grpc-middleware@v1.4.0/chain.go:86 +0x218
  google.golang.org/grpc.(*Server).processStreamingRPC()
      /go/pkg/mod/google.golang.org/grpc@v1.62.1/server.go:1687 +0x1f50
  google.golang.org/grpc.(*Server).handleStream()
      /go/pkg/mod/google.golang.org/grpc@v1.62.1/server.go:1801 +0x19f3
  google.golang.org/grpc.(*Server).serveStreams.func2.1()
      /go/pkg/mod/google.golang.org/grpc@v1.62.1/server.go:1027 +0x15c

Goroutine 2253766 (running) created at:
  github.com/tikv/pd/pkg/schedule.(*Coordinator).Run()
      /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/pd/pkg/schedule/coordinator.go:491 +0x44a
  github.com/tikv/pd/pkg/schedule.(*Coordinator).RunUntilStop()
      /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/pd/pkg/schedule/coordinator.go:455 +0x48
  github.com/tikv/pd/server/cluster.(*schedulingController).runCoordinator()
      /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/pd/server/cluster/scheduling_controller.go:121 +0xe6
  github.com/tikv/pd/server/cluster.(*schedulingController).startSchedulingJobs.func2()
      /home/jenkins/agent/workspace/build-common/go/src/github.com/pingcap/pd/server/cluster/scheduling_controller.go:98 +0x33

Goroutine 37567 (running) created at:
  google.golang.org/grpc.(*Server).serveStreams.func2()
      /go/pkg/mod/google.golang.org/grpc@v1.62.1/server.go:1038 +0x224
  google.golang.org/grpc/internal/transport.(*http2Server).operateHeaders()
      /go/pkg/mod/google.golang.org/grpc@v1.62.1/internal/transport/http2_server.go:619 +0x3a01
  google.golang.org/grpc/internal/transport.(*http2Server).HandleStreams()
      /go/pkg/mod/google.golang.org/grpc@v1.62.1/internal/transport/http2_server.go:660 +0x285
  google.golang.org/grpc.(*Server).serveStreams()
      /go/pkg/mod/google.golang.org/grpc@v1.62.1/server.go:1021 +0x6bb
  google.golang.org/grpc.(*Server).handleRawConn.func1()
      /go/pkg/mod/google.golang.org/grpc@v1.62.1/server.go:957 +0x86
==================
```

### PR DESCRIPTION
**Pull request by lhy1024**
### What problem does this PR solve?

Issue Number: Close #8263

### What is changed and how does it work?

```commit-message
```

### Check List

Tests

- Unit test
- Manual test (add detailed scripts or steps below)
`go test -timeout 120s -run ^TestOperatorCheckConcurrently$ github.com/tikv/pd/pkg/schedule/operator -race`
- in master
```
==================
WARNING: DATA RACE
Read at 0x00c0004dc240 by goroutine 212:
  github.com/tikv/pd/pkg/schedule/operator.(*Operator).Check()
      /home/lhy1024/pd/pkg/schedule/operator/operator.go:382 +0x5c4
  github.com/tikv/pd/pkg/schedule/operator.TestOperatorCheckConcurrently.func1()
      /home/lhy1024/pd/pkg/schedule/operator/operator_test.go:593 +0xbe

Previous write at 0x00c0004dc240 by goroutine 215:
  sync/atomic.CompareAndSwapInt64()
      /usr/lib/go/src/runtime/race_amd64.s:316 +0xb
  sync/atomic.CompareAndSwapInt64()
      <autogenerated>:1 +0x1a
  github.com/tikv/pd/pkg/schedule/operator.TestOperatorCheckConcurrently.func1()
      /home/lhy1024/pd/pkg/schedule/operator/operator_test.go:593 +0xbe

Goroutine 212 (running) created at:
  github.com/tikv/pd/pkg/schedule/operator.TestOperatorCheckConcurrently()
      /home/lhy1024/pd/pkg/schedule/operator/operator_test.go:591 +0x487
  testing.tRunner()
      /usr/lib/go/src/testing/testing.go:1689 +0x21e
  testing.(*T).Run.gowrap1()
      /usr/lib/go/src/testing/testing.go:1742 +0x44

Goroutine 215 (running) created at:
  github.com/tikv/pd/pkg/schedule/operator.TestOperatorCheckConcurrently()
      /home/lhy1024/pd/pkg/schedule/operator/operator_test.go:591 +0x487
  testing.tRunner()
      /usr/lib/go/src/testing/testing.go:1689 +0x21e
  testing.(*T).Run.gowrap1()
      /usr/lib/go/src/testing/testing.go:1742 +0x44
==================
--- FAIL: TestOperatorCheckConcurrently (0.00s)
    testing.go:1398: race detected during execution of test
FAIL
FAIL    github.com/tikv/pd/pkg/schedule/operator        0.103s
FAIL
```
- with this pr
```
ok      github.com/tikv/pd/pkg/schedule/operator        1.098s
```

### Release note

```release-note
None.
```

### CODE REVIEW

File: pkg/schedule/operator/operator.go [hunk 1]
```diff
@@ -376,10 +376,11 @@ func (o *Operator) Check(region *core.RegionInfo) OpStep {
 	defer func() { _ = o.CheckTimeout() }()
 	for step := atomic.LoadInt32(&o.currentStep); int(step) < len(o.steps); step++ {
 		if o.steps[int(step)].IsFinish(region) {
-			if atomic.CompareAndSwapInt64(&(o.stepsTime[step]), 0, time.Now().UnixNano()) {
+			current := time.Now()
+			if atomic.CompareAndSwapInt64(&(o.stepsTime[step]), 0, current.UnixNano()) {
 				startTime, _ := o.getCurrentTimeAndStep()
 				operatorStepDuration.WithLabelValues(reflect.TypeOf(o.steps[int(step)]).Name()).
-					Observe(time.Unix(0, o.stepsTime[step]).Sub(startTime).Seconds())
```
Comment (rleungx) [inline pkg/schedule/operator/operator.go:382]: Is stepsTime useless?

Comment (lhy1024) [reply pkg/schedule/operator/operator.go:382]: If the code goes into this branch, it means that `o.stepsTime[step])` is now `time.Now()`

**Review by rleungx — APPROVED** [review]

**Review by okJiang — APPROVED** [review]

**Review by HuSharp — APPROVED** [review]

**Comment by HuSharp** [comment]
please add affect label

**Pull request merged**